### PR TITLE
42 fe create new animal

### PR DESCRIPTION
**Pull request by s2an**
Enormous PR here.

Created (for animals & shelters) their required poros, services, facades, controllers, routes, views, and respective tests.

I tried to fill out everything as much as possible.

The fixture file (for shelters) is from seeded data into the backend (only on my local), due to the connection not being set up yet.

The poros required a require_relative "../../app/poros/animal" and I don't know why the connection wasn't automatic.

I added notes where possible in the code to highlight what is needed or needs to be updated.

I hope the merge conflicts are not too bad.

(I did need to manually remove some coverage links, so hopefully all of that is broken and deleted from the cache.

Not everything passes, but I tried to get it as close to done as possible.

Isaac was pivotal in getting this setup and as far as it did get, but all mistakes are mine 😉

### CODE REVIEW

File: app/controllers/animals_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  private
+
+  # this heavyweight method allows for a collection dropdown in the view
+  def allowed_species
+    [
+      { id: 1, name: "Chicken" },
+      { id: 2, name: "Bee" },
+    ].map { |species| OpenStruct.new(species) }
+  end
+end
```
Comment (megroth2) [inline app/controllers/animals_controller.rb:23]: oooo this is cool

**Review by tmitchellisaac — COMMENTED** [review]
def get_shelter(shelter_id)
    get_url("api/v1/shelters/#{shelter_id}")
  end
end
wondering now if this should be
"api/v1/users/#{user_id/shelters/#{shelter_id}"
(edited)

File: spec/services/shelter_service_spec.rb [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
-# require "pry"; binding.pry
-    expect(shelter).to have_key(:name)
-    expect(shelter[:name]).to be_a(String)
-    expect(shelter[:name]).to eq("red barn")
+  describe "shelter service" do
+    it "calls a shelter" do
+      # @shelter = Shelter.new(name: "red barn")
+
+      expect(shelter).to have_key(:name)
+      expect(shelter[:name]).to be_a(String)
```
Comment (tmitchellisaac) [inline spec/services/shelter_service_spec.rb:9]: might have to add an actual `shelter` so it knows what we're talking about

File: app/services/animal_service.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    end
+  end
+
+  def get_url(url)
+    response = conn.get(url)
+    JSON.parse(response.body, symbolize_names: true)
+  end
+
+  def get_animal(animal_id)
+    get_url("api/v1/animals/#{animal_id}")
```
Comment (tmitchellisaac) [inline app/services/animal_service.rb:13]: we'll have to get together to talk about how these requests are gonna look. like if we should include the user/shelters in the animal request. I honestly don't know which one would be better or if it matters

File: spec/fixtures/shelter_1.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  "attributes": {
+  "name": "red barn"
+  },
+  "relationships": {
+  "animals": {
+  "data": []
+  }
+  }
+  }
+  }
```
Comment (megroth2) [inline spec/fixtures/shelter_1.json:14]: could we fix the indentations here? Is this right?
```
{
  "data": {
    "id": "1",
    "type": "shelter",
    "attributes": {
      "name": "red barn"
    },
    "relationships": {
      "animals": {
        "data": []
      }
    }
  }
}
```

File: spec/features/animals/new_animal_spec.rb [hunk 5]
```diff
@@ -38,29 +37,29 @@
...
-      expect(current_path).to eq(shelter_animal(animal))
+      expect(current_path).to eq("/shelters/1/animals/1")
       
       # user-defined attributes
       expect(page).to have_content("Animal successfully created.")
-      expect(page).to have_content("chicken")
       expect(page).to have_content("Mickey McCluckkiddy")
+      expect(page).to have_content("Chicken")
+      expect(page).to have_content("03/03/2020")
       expect(page).to have_content("black with orange spots")
```
Comment (tmitchellisaac) [inline spec/features/animals/new_animal_spec.rb:55]: does this pass? if so thats sick!

File: spec/poros/animal_spec.rb [hunk 6]
```diff
@@ -1,5 +1,5 @@
 require "spec_helper"
-require_relative "../../app/poros/animal"
+require_relative "../../app/poros/animal" # <-- I don't know why this is needed
```
Comment (tmitchellisaac) [inline spec/poros/animal_spec.rb:3]: yeah idk about that at all really

File: app/views/animals/new.html.erb [hunk 7]
```diff
@@ -2,7 +2,8 @@
 <hr>
 <h2>New Animal Form</h2>
 <div class="new_animal_form">
-  <%= form_with url: "/shelters/#{@facade.shelter.id}/animals/#{animal.id}", method: :post, data: {turbo: false} do |form| %>
+  <%# the form_with url may not be correct %>
+  <%= form_with url: "/shelters/#{@facade.shelter.id}/animals/#{@facade.animal.id}", method: :post, data: {turbo: false} do |form| %>
```
Comment (tmitchellisaac) [inline app/views/animals/new.html.erb:6]: it would be `"/animals/create"` right?? I'm blanking

**Comment by tmitchellisaac** [comment]
I'm cool to merge this. I think it makes enough sense to me, that if we switched roles here soon i'd be able to figure out whats up

File: spec/services/shelter_service_spec.rb (hunk 2 above)
Comment (s2an) [reply spec/services/shelter_service_spec.rb:9]: Should be able to use the service to get a shelter once the API is linked correctly.

File: app/services/animal_service.rb (hunk 3 above)
Comment (s2an) [reply app/services/animal_service.rb:13]: Good point! I did not think about that.

File: spec/fixtures/shelter_1.json (hunk 4 above)
Comment (s2an) [reply spec/fixtures/shelter_1.json:14]: That looks correct. That fixture was temporary (just pulled from seed data on my back-end's local), just to get something to show it was working. Once a real fixture is in place, the tests will use that.

File: app/views/animals/new.html.erb (hunk 7 above)
Comment (s2an) [reply app/views/animals/new.html.erb:6]: I believe you are correct!

**Pull request merged**